src/streaming.ts

```diff
@@ -46,46 +51,116 @@ export class StreamProcessor {
     };
 
-    eventSource.onretrying = (event: MessageEvent) => {
-      this.log.error('Stream is trying to reconnect', event);
-      // Only inform the client once that we're retrying
-      if (!this.isRetrying) {
-        this.isRetrying = true
-        this.eventBus.emit(StreamEvent.RETRYING);
-      }
-    };
+    const failed = (msg) => {
+      const delayMs = this.getRandomRetryDelayMs();
+      this.log.warn(`SSE disconnected: ${msg} will retry in ${delayMs}ms`);
+      this.readyState = StreamProcessor.RETRYING;
+      this.eventBus.emit(StreamEvent.RETRYING);
 
-    eventSource.onerror = (event: MessageEvent) => {
-      this.log.error('Unrecoverable error with stream, not retrying to connect: ', event);
-      this.eventBus.emit(StreamEvent.ERROR, event);
+      setTimeout(() => {
+        this.log.info('SSE retrying to connect');
+        this.connect(url, options, connected, failed);
+      }, delayMs);
     };
 
-    eventSource.addEventListener('*', (event: MessageEvent) => {
-      const msg: StreamMsg = JSON.parse(event.data);
+    this.connect(url, options, connected, failed);
 
-      this.log.debug('Received event from stream: ', msg);
+    this.eventBus.emit(StreamEvent.READY);
+  }
 
+  private getRandomRetryDelayMs(): number {
+    const delayMs = this.retryDelayMs * this.retryAttempt;
+    this.retryAttempt += 1;
+    return Math.min(delayMs, 60000);
+  }
+
+  private connect(
+    url: string,
+    options: RequestOptions,
+    connected,
+    failed,
+  ): void {
+    if (this.readyState === StreamProcessor.CONNECTED) {
+      this.log.debug('SSE already connected, skip retry');
+      return;
+    }
+
+    const isSecure = url.startsWith('https:');
+    this.log.debug('SSE HTTP start request', url);
+
+    (isSecure ? https : http)
+      .request(url, options, (res) => {
+        this.log.debug('SSE got HTTP response code', res.statusCode);
+
+        if (res.statusCode !== 200) {
+          failed('HTTP code ' + res.statusCode);
+          return;
+        }
+
+        connected();
+
+        res
+          .on('data', (data) => {
+            this.processData(data);
+          })
+          .on('close', () => {
+            failed('SSE stream closed');
+          });
+        //.on('end', ()=> { failed('SSE stream ended'); })
+      })
+      .on('error', (err) => {
+        failed('SSE request failed ' + err.message);
+      })
+      .on('timeout', () => {
+        failed(
+          'SSE request timed out after ' +
+            StreamProcessor.SSE_TIMEOUT_MS +
+            'ms',
+        );
+      })
+      .setTimeout(StreamProcessor.SSE_TIMEOUT_MS)
+      .end();
+  }
+
+  private processData(data: any): void {
+    const lines = data.toString().split(/\r?\n/);
+    lines.forEach((line) => this.processLine(line));
+  }
+
+  private processLine(line: string): void {
+    if (line.startsWith('data:')) {
+      this.log.debug('SSE GOT:', line.substring(5));
+      const msg = JSON.parse(line.substring(5));
```

Comment:
Minor thing, but just to keep things readable, it's best to keep a bit of white space before conditions.
```suggestion
      const msg = JSON.parse(line.substring(5));

```